src/workerd/api/encoding.c++

```diff
@@ -460,38 +464,584 @@ kj::Maybe<jsg::JsString> TextDecoder::decodePtr(
+      KJ_DASSERT(!(isLeadSurrogate(c) && inputPos + 1 < input.size() &&
+                     isTrailSurrogate(input[inputPos + 1])),
+          "Valid surrogate pair should not trigger SURROGATE error");
+
+      // Invalid surrogate = U+FFFD (3 bytes)
+      utf8Length += 3;
+      inputPos++;
+    } else {
+      KJ_FAIL_REQUIRE(
+          "Unexpected UTF-16 validation error from simdutf", static_cast<int>(result.error));
+    }
+  }
+
+  return utf8Length;
+}
+
+// Encode a single UTF-16 code unit to UTF-8
+inline size_t encodeUtf8CodeUnit(char16_t c, kj::ArrayPtr<char> out) {
+  if (c < 0x80) {
+    KJ_DASSERT(out.size() >= 1);
+    out[0] = static_cast<char>(c);
+    return 1;
+  } else if (c < 0x800) {
+    KJ_DASSERT(out.size() >= 2);
+    out[0] = static_cast<char>(0xC0 | (c >> 6));
+    out[1] = static_cast<char>(0x80 | (c & 0x3F));
+    return 2;
+  } else {
+    KJ_DASSERT(out.size() >= 3);
+    out[0] = static_cast<char>(0xE0 | (c >> 12));
+    out[1] = static_cast<char>(0x80 | ((c >> 6) & 0x3F));
+    out[2] = static_cast<char>(0x80 | (c & 0x3F));
+    return 3;
+  }
+}
+
+// Convert UTF-16 with potentially invalid surrogates to UTF-8.
+// Invalid surrogates are replaced with U+FFFD.
+// Returns the number of UTF-8 bytes written.
+// Uses SIMD for valid portions and falls back to scalar for invalid surrogates.
+size_t convertInvalidUtf16ToUtf8(kj::ArrayPtr<const char16_t> input, kj::ArrayPtr<char> out) {
+  size_t inputPos = 0;
+  size_t outputPos = 0;
+
+  while (inputPos < input.size()) {
+    // Find the next invalid surrogate using SIMD validation
+    auto result =
+        simdutf::validate_utf16_with_errors(input.begin() + inputPos, input.size() - inputPos);
+
+    if (result.error == simdutf::error_code::SUCCESS) {
+      // Remaining input is valid - convert it all with SIMD
+      outputPos += simdutf::convert_utf16_to_utf8(
+          input.begin() + inputPos, input.size() - inputPos, out.begin() + outputPos);
+      KJ_DASSERT(outputPos <= out.size());
+      break;
+    }
+
+    if (result.error == simdutf::error_code::SURROGATE) {
+      // Convert the valid portion before the error with SIMD
+      if (result.count > 0) {
+        outputPos += simdutf::convert_valid_utf16_to_utf8(
+            input.begin() + inputPos, result.count, out.begin() + outputPos);
+        KJ_DASSERT(outputPos <= out.size());
+        inputPos += result.count;
+      }
+
+      // Handle the invalid surrogate at inputPos
+      // SURROGATE error means unpaired surrogate, so valid pair should be impossible
+      [[maybe_unused]] char16_t c = input[inputPos];
+      KJ_DASSERT(!(isLeadSurrogate(c) && inputPos + 1 < input.size() &&
+                     isTrailSurrogate(input[inputPos + 1])),
+          "Valid surrogate pair should not trigger SURROGATE error");
+
+      // Invalid surrogate - replace with U+FFFD (3 bytes)
+      outputPos += encodeUtf8CodeUnit(0xFFFD, out.slice(outputPos, out.size()));
```

Comment:
The UTF-8 encoding of U+FFFD is a known value; you could theoretically avoid calculating it every time.